## Assignment 07
### Note:
- For visualization, you have the freedom to choose one or many from
    - Matplotlib
    - Seaborn
    - Plotly Express 
    - any other library
- You can use the local Jupyter environment if you are not comfortable with Colab for this assignment.
- Use use this notebook as your template and follow the instructions

The first half of this assignment is a repeat of assignment 05.

This gives you a chance to refresh.

The second half is new and gives you a chance to perform additional practices.

You also get a chance to use some of the Python libraries and techniques.

You will unzip the zip file manually in preparation for this assignment.

The links to the zip file is (you have already used it in your assignment 5):

- https://data.ed.gov/dataset/college-scorecard-all-data-files-through-6-2020/resources (The page containing link to the zip file)

- https://ed-public-download.app.cloud.gov/downloads/CollegeScorecard_Raw_Data_01192021.zip (The link to download the zip file)

After unzipping, You will have all files in one folder named "CollegeScorecard_Raw_Data" 

The folder contains the yearly data files from 1996 to 2019 school years.

if you are using your local environment, I suggest you create a new folder call "Assignment7" and place the Notebook file and the data folder under it.
```
Assignment7/
    - assignment_07.ipynb
    - CollegeScorecard_Raw_Data/
        - ...
        - 'MERGED1996_97_PP.csv',
        - 'MERGED2015_16_PP.csv',
        - ...
        - 'MERGED2017_18_PP.csv'
```


In [1]:
# <1> 
# Import libraries

#(Write code here)

import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
import csv
import os

In [2]:
# <2> 
# Display the current workin directory using os.getcwd()
# you would need to import a standard Python library called os which stands for operating system,
# so place that import statement in the previous cell.
# Since your notebook and your data files are not in # the same folder, 
# you want to make sure what is the current working folder and 
# how to access a data file in a different folder

#(Write code here)

os.getcwd()

'/content'

In [3]:
%cd /content/Assignment-7

/content/Assignment-7


In [4]:
# <3> 
# os library has a method call listdir which generates a list of files in a directory/folder.
# use this method to assign the contents (list of file names) of the data folder to a variable and display it.
# # You can use ../ construct to traverse to the parent folder and then to another foloder parallel to the current folder

#(Write code here)

files = os.listdir()
files

['MERGED2008_09_PP.csv',
 'MERGED2002_03_PP.csv',
 'MERGED2004_05_PP.csv',
 'MERGED2009_10_PP.csv',
 'MERGED2016_17_PP.csv',
 'MERGED2001_02_PP.csv',
 'MERGED2005_06_PP.csv',
 'MERGED1999_00_PP.csv',
 'MERGED2003_04_PP.csv',
 'MERGED2000_01_PP.csv',
 'MERGED2010_11_PP.csv',
 'MERGED2015_16_PP.csv',
 'MERGED2006_07_PP.csv',
 'MERGED2014_15_PP.csv',
 'MERGED1997_98_PP.csv',
 'MERGED2011_12_PP.csv',
 'MERGED1996_97_PP.csv',
 'MERGED2013_14_PP.csv',
 'MERGED2012_13_PP.csv',
 'MERGED2018_19_PP.csv',
 'MERGED2017_18_PP.csv',
 'MERGED2007_08_PP.csv',
 'MERGED1998_99_PP.csv']

In [5]:
# <4> 
# The folder contains files that are not the yearly data files
# write code to remove the unwanted files from the list vsriable 
# Note: don't remove/delete these files from the folder in your drive
# For example, use the file extension to only keep the csv files 
# or use the name patter - data file name begins with "MERGED"
# You can use the concept of list comprehension to write just 
# one line of code as well as using a for loop, your choice.

#(Write code here)

path ="/content/Assignment-7"

all_csv = []

for file in os.listdir(path):
  if file.startswith("MERGED"):
    all_csv.append(file)
all_csv

['MERGED2008_09_PP.csv',
 'MERGED2002_03_PP.csv',
 'MERGED2004_05_PP.csv',
 'MERGED2009_10_PP.csv',
 'MERGED2016_17_PP.csv',
 'MERGED2001_02_PP.csv',
 'MERGED2005_06_PP.csv',
 'MERGED1999_00_PP.csv',
 'MERGED2003_04_PP.csv',
 'MERGED2000_01_PP.csv',
 'MERGED2010_11_PP.csv',
 'MERGED2015_16_PP.csv',
 'MERGED2006_07_PP.csv',
 'MERGED2014_15_PP.csv',
 'MERGED1997_98_PP.csv',
 'MERGED2011_12_PP.csv',
 'MERGED1996_97_PP.csv',
 'MERGED2013_14_PP.csv',
 'MERGED2012_13_PP.csv',
 'MERGED2018_19_PP.csv',
 'MERGED2017_18_PP.csv',
 'MERGED2007_08_PP.csv',
 'MERGED1998_99_PP.csv']

In [6]:
# <5> 
# Now that you have a clean list of the yearly files, you want to loop through them
# and read them into a dataframe one at a time. You only load six columns: 
# ["UNITID", "INSTNM", "STABBR", "REGION", "ADM_RATE", "TUITIONFEE_IN"].
# You should use "usecols" option of Pandas to avoid reading unwanted columns.
# You also want to add a new column call "YEAR" to differentiate the data frames from each other.
# The YEAR variable should be yyyy format so tht you can convert them into integer.
# if you use the format yyyy-yy (such as 1997-98 school year), you will not be able to convert them directly to integer
# if you use scatter plot, the YEAR needs to be converted to integer or float. 
# 
# You would use an empty list and append the yearly dataframes to the list.
# After all data files are loaded and appended to the list, you would use Pandas to concatenate them into a 
# new single data frame.
# Note: this exercise incorporates many techques we learned before
# - list (creating an empty, append an item to the list)
# - for loop 
# - read only the needed columns from a file (using usecols option)
# - add a new column to a data frame
# - concatenate multiple dataframes into a single one
# This exercise may appear a big challenging but it worths the effort. You will learn a lot and  love it. I promise.

#(Write code here)

all_df_list = []

for csv in all_csv:
    df = pd.read_csv(csv, usecols = ["UNITID", "INSTNM", "STABBR", "REGION", "ADM_RATE", "TUITIONFEE_IN"])
    Year = csv[6:10]
    df['Year'] = Year
    all_df_list.append(df)

final_df = pd.concat(all_df_list)
final_df

UNITID                                       INSTNM  ... TUITIONFEE_IN  Year
0     100636           Community College of the Air Force  ...           NaN  2008
1     100654                     Alabama A & M University  ...        4072.0  2008
2     100663          University of Alabama at Birmingham  ...        4664.0  2008
3     100690                           Amridge University  ...        6600.0  2008
4     100706          University of Alabama in Huntsville  ...        5952.0  2008
...      ...                                          ...  ...           ...   ...
6697  434779                  WATERMAN TRAINING INSTITUTE  ...           NaN  1998
6698  434849                                CSC Institute  ...           NaN  1998
6699  434867    CENTER FOR EMPLOYMENT TRAINING-SANTA ROSA  ...           NaN  1998
6700  434876  CENTER FOR EMPLOYMENT TRAINING-TRACY CAMPUS  ...           NaN  1998
6701  434900                    Dewey University-Carolina  ...           NaN  1998

[163331 rows x 7 columns]

In [7]:
# <6> 
# explore the new dataframe (# of observations, varibles, head, tail, sample, missing values, statistics,etc.)

#(Write code here)

final_df.shape

(163331, 7)

In [8]:
final_df.head()

UNITID                               INSTNM  ... TUITIONFEE_IN  Year
0  100636   Community College of the Air Force  ...           NaN  2008
1  100654             Alabama A & M University  ...        4072.0  2008
2  100663  University of Alabama at Birmingham  ...        4664.0  2008
3  100690                   Amridge University  ...        6600.0  2008
4  100706  University of Alabama in Huntsville  ...        5952.0  2008

[5 rows x 7 columns]

In [9]:
final_df.tail()

UNITID                                       INSTNM  ... TUITIONFEE_IN  Year
6697  434779                  WATERMAN TRAINING INSTITUTE  ...           NaN  1998
6698  434849                                CSC Institute  ...           NaN  1998
6699  434867    CENTER FOR EMPLOYMENT TRAINING-SANTA ROSA  ...           NaN  1998
6700  434876  CENTER FOR EMPLOYMENT TRAINING-TRACY CAMPUS  ...           NaN  1998
6701  434900                    Dewey University-Carolina  ...           NaN  1998

[5 rows x 7 columns]

In [10]:
final_df.sample()

UNITID                                  INSTNM  ... TUITIONFEE_IN  Year
2060  165255  Catherine Hinds Institute of Esthetics  ...           NaN  1998

[1 rows x 7 columns]

In [11]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 163331 entries, 0 to 6701
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   UNITID         163331 non-null  int64  
 1   INSTNM         163331 non-null  object 
 2   STABBR         163331 non-null  object 
 3   REGION         163330 non-null  float64
 4   ADM_RATE       44912 non-null   float64
 5   TUITIONFEE_IN  79195 non-null   float64
 6   Year           163331 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 10.0+ MB


In [12]:
final_df.describe()

UNITID         REGION      ADM_RATE  TUITIONFEE_IN
count  1.633310e+05  163330.000000  44912.000000   79195.000000
mean   9.432395e+05       4.554693      0.692951   11665.636871
std    4.571721e+06       2.189217      0.217058    9986.003542
min    1.006360e+05       0.000000      0.000000       0.000000
25%    1.633380e+05       3.000000      0.555075    3776.000000
50%    2.127800e+05       5.000000      0.717500    9285.000000
75%    4.075680e+05       6.000000      0.857100   16148.000000
max    4.914640e+07       9.000000      1.000000   74787.000000

In [13]:
pd.isnull(final_df)           # checking for null values in every column

UNITID  INSTNM  STABBR  REGION  ADM_RATE  TUITIONFEE_IN   Year
0      False   False   False   False      True           True  False
1      False   False   False   False     False          False  False
2      False   False   False   False     False          False  False
3      False   False   False   False      True          False  False
4      False   False   False   False     False          False  False
...      ...     ...     ...     ...       ...            ...    ...
6697   False   False   False   False      True           True  False
6698   False   False   False   False      True           True  False
6699   False   False   False   False      True           True  False
6700   False   False   False   False      True           True  False
6701   False   False   False   False      True           True  False

[163331 rows x 7 columns]

#### isnull() returns a boolean value where true represents a null value and false represents that there is a value

In [14]:
final_df.isnull().values.any()      # another way to just check if there are any null values in the dataframe as a whole

True

In [15]:
final_df.isnull().sum()

UNITID                0
INSTNM                0
STABBR                0
REGION                1
ADM_RATE         118419
TUITIONFEE_IN     84136
Year                  0
dtype: int64

#### This shows that there are 118419 null values in the column 'ADM_RATE' and there are 84136 null values in 'TUITIONFEE_IN' column

In [16]:
final_df.isnull().sum().sum()       # counts the total number of null vales in the dataframe

202556

In [17]:
final_df.columns        # to get the column names of the dataframe

Index(['UNITID', 'INSTNM', 'STABBR', 'REGION', 'ADM_RATE', 'TUITIONFEE_IN',
       'Year'],
      dtype='object')

In [18]:
final_df

UNITID                                       INSTNM  ... TUITIONFEE_IN  Year
0     100636           Community College of the Air Force  ...           NaN  2008
1     100654                     Alabama A & M University  ...        4072.0  2008
2     100663          University of Alabama at Birmingham  ...        4664.0  2008
3     100690                           Amridge University  ...        6600.0  2008
4     100706          University of Alabama in Huntsville  ...        5952.0  2008
...      ...                                          ...  ...           ...   ...
6697  434779                  WATERMAN TRAINING INSTITUTE  ...           NaN  1998
6698  434849                                CSC Institute  ...           NaN  1998
6699  434867    CENTER FOR EMPLOYMENT TRAINING-SANTA ROSA  ...           NaN  1998
6700  434876  CENTER FOR EMPLOYMENT TRAINING-TRACY CAMPUS  ...           NaN  1998
6701  434900                    Dewey University-Carolina  ...           NaN  1998

[163331 rows x 7 columns]

In [19]:
# <7>
# the dataframe contains many years of data of all U.S. colleges.
# let's just look at UMBC
# filter/query the dataframe to retrieve only rows that belong to UMBC
# save the UMNC data to a new data frame. using a new variable 
# so that the old big data frame is still available for later use.

#(Write code here)

final_df[final_df['INSTNM'].str.contains("University of Maryland-Baltimore County")]   # getting rows that have UMBC in institute name 

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1915  163268  University of Maryland-Baltimore County  ...        5910.0  2001
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1873  163268  University of Maryland-Baltimore County  ...        7388.0  2003
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
1710  163268  University of Maryland-Baltimore County  ...        9171.0  2010
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1799  163268  University of Maryland-Baltimore County  ...        8622.0  2006
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
1629  163268  University of Maryland-Baltimore County  ...       10068.0  2013
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998

[23 rows x 7 columns]

In [20]:
UMBC_df = final_df[final_df['INSTNM'].str.contains("University of Maryland-Baltimore County")]      # saving UMBC data in new dataframe

In [21]:
# <8>
# Explore this UMBC dataframe (# of observations, varibles, head, tail, sample, missing values, statistics,etc.)

#(Write code here)

UMBC_df

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1915  163268  University of Maryland-Baltimore County  ...        5910.0  2001
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1873  163268  University of Maryland-Baltimore County  ...        7388.0  2003
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
1710  163268  University of Maryland-Baltimore County  ...        9171.0  2010
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1799  163268  University of Maryland-Baltimore County  ...        8622.0  2006
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
1629  163268  University of Maryland-Baltimore County  ...       10068.0  2013
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998

[23 rows x 7 columns]

In [22]:
UMBC_df.shape

(23, 7)

In [23]:
UMBC_df.columns             # to find the columns of UMBC dataframe

Index(['UNITID', 'INSTNM', 'STABBR', 'REGION', 'ADM_RATE', 'TUITIONFEE_IN',
       'Year'],
      dtype='object')

In [24]:
UMBC_df.head()                 # to find the first five rows of UMBC dataframe

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016

[5 rows x 7 columns]

In [25]:
UMBC_df.tail()          # to find the last five rows of UMBC dataframe

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998

[5 rows x 7 columns]

In [26]:
UMBC_df.sample()

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002

[1 rows x 7 columns]

In [27]:
UMBC_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23 entries, 1767 to 1982
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   UNITID         23 non-null     int64  
 1   INSTNM         23 non-null     object 
 2   STABBR         23 non-null     object 
 3   REGION         23 non-null     float64
 4   ADM_RATE       18 non-null     float64
 5   TUITIONFEE_IN  19 non-null     float64
 6   Year           23 non-null     object 
dtypes: float64(3), int64(1), object(3)
memory usage: 1.4+ KB


In [28]:
UMBC_df.describe()

UNITID  REGION   ADM_RATE  TUITIONFEE_IN
count      23.0    23.0  18.000000      19.000000
mean   163268.0     2.0   0.640694    9004.842105
std         0.0     0.0   0.052728    1825.795421
min    163268.0     2.0   0.568300    5490.000000
25%    163268.0     2.0   0.597600    8270.000000
50%    163268.0     2.0   0.632100    8872.000000
75%    163268.0     2.0   0.689375   10226.000000
max    163268.0     2.0   0.717400   11778.000000

In [29]:
UMBC_df.isnull().values.any()     # to find if there are any null values in the entire dataframe

True

In [30]:
pd.isnull(UMBC_df)                # to check which values are null values

UNITID  INSTNM  STABBR  REGION  ADM_RATE  TUITIONFEE_IN   Year
1767   False   False   False   False     False          False  False
1893   False   False   False   False     False          False  False
1830   False   False   False   False     False          False  False
1745   False   False   False   False     False          False  False
1510   False   False   False   False     False          False  False
1915   False   False   False   False     False          False  False
1819   False   False   False   False     False          False  False
1955   False   False   False   False      True           True  False
1873   False   False   False   False     False          False  False
1926   False   False   False   False      True          False  False
1710   False   False   False   False     False          False  False
1555   False   False   False   False     False          False  False
1799   False   False   False   False     False          False  False
1610   False   False   False   False     False          False  False
2049   False   False   False   False      True           True  False
1694   False   False   False   False     False          False  False
2151   False   False   False   False      True           True  False
1629   False   False   False   False     False          False  False
1660   False   False   False   False     False          False  False
1415   False   False   False   False     False          False  False
1478   False   False   False   False     False          False  False
1779   False   False   False   False     False          False  False
1982   False   False   False   False      True           True  False

#### In the above output 'True' represents the null value and 'False' represents that there is a value

In [31]:
UMBC_df.isnull().sum()

UNITID           0
INSTNM           0
STABBR           0
REGION           0
ADM_RATE         5
TUITIONFEE_IN    4
Year             0
dtype: int64

#### The above output represents that there are 5 null values in the column 'ADM_RATE' and there are 4 null values in the column 'TUITIONFEE_IN'

In [32]:
UMBC_df.isnull().sum().sum()

9

#### The above output represents that there are 9 null values in the entire UMBC dataframe

In [33]:
UMBC_df

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1915  163268  University of Maryland-Baltimore County  ...        5910.0  2001
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1873  163268  University of Maryland-Baltimore County  ...        7388.0  2003
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
1710  163268  University of Maryland-Baltimore County  ...        9171.0  2010
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1799  163268  University of Maryland-Baltimore County  ...        8622.0  2006
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
1629  163268  University of Maryland-Baltimore County  ...       10068.0  2013
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998

[23 rows x 7 columns]

In [34]:
UMBC_df = UMBC_df.sort_values(by=['Year'])        # sorting the dataframe as per the year column

In [35]:
UMBC_df

UNITID                                   INSTNM  ... TUITIONFEE_IN  Year
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
1915  163268  University of Maryland-Baltimore County  ...        5910.0  2001
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1873  163268  University of Maryland-Baltimore County  ...        7388.0  2003
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005
1799  163268  University of Maryland-Baltimore County  ...        8622.0  2006
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1710  163268  University of Maryland-Baltimore County  ...        9171.0  2010
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1629  163268  University of Maryland-Baltimore County  ...       10068.0  2013
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018

[23 rows x 7 columns]

In [36]:
# <9> 
# Plot UMBC's in-state tution overtime from 1996 to 2019 using Bar Chart

#(Write code here)

fig = px.bar(UMBC_df,
             x = "Year",
             y = "TUITIONFEE_IN",
             title = "In-state Tution fee as per year (Bar chart)",
             color = "TUITIONFEE_IN",
             height = 600)             
            
fig.show()

In [37]:
# <10> 
# Plot UMBC's in-state tution overtime from 1996 to 2019 using Line Chart

#(Write code here)

fig = px.line(UMBC_df,
             x = "Year",
             y = "TUITIONFEE_IN",
             title = "In-state Tution fee as per year (Line chart)")
            
fig.show()

In [38]:
# <11>
# Now let's look at the tuition growth rate year over year
# We need to calculate UMBC tuition change percentage each year
# First convert the TUITIONFEE_IN column to a Python List

#(Write code here)

fee_list = UMBC_df["TUITIONFEE_IN"].tolist()        # converting column in dataframe to python list
fee_list

[nan,
 nan,
 nan,
 nan,
 5490.0,
 5910.0,
 6362.0,
 7388.0,
 8020.0,
 8520.0,
 8622.0,
 8708.0,
 8780.0,
 8872.0,
 9171.0,
 9467.0,
 9764.0,
 10068.0,
 10384.0,
 11006.0,
 11264.0,
 11518.0,
 11778.0]

#### From the above we can say that we have tution fee from the year 2000 and fee for the previous years was given as null values and we can not do anything with that. And the data ended with the year 2018. Hence calculating the tution fee change percentage from 2000 to 2018.

In [39]:
# Loop through the list and calculate the % change each year over the prior year
# This takes some effort. Not hard, just some abstract/logical thinking and some experiments
# Have fun on this one

#(Write code here)



In [40]:
# The resulting number has many decimal points which are unnecessary and not visually appealing.
# Round up the percentage to two decimal points
# You can use for loop. Or better, use list comprehension for simplicity/brevity

#(Write code here)

In [41]:
# Add the list of the percentages to the dataframe as a new column ("PCT_CHANGE")
# Not as hard as you may think. If you get stuck, you think too hard. 
# Google it and you will find the answer.

#(Write code here)

In [42]:
UMBC_pct_chng = []

for i in range(0, len(fee_list)):
  if i == 0:                           # no previous data to compare the first data
    UMBC_pct_chng.append(np.nan)
  else:
    UMBC_pct_chng.append(((fee_list[i] - fee_list[i - 1]) / fee_list[i - 1]) * 100)     # finding percentage change

UMBC_pct_chng = [round(i, 2) for i in UMBC_pct_chng]          # rounding up the value

UMBC_df["PCT_CHANGE"] = UMBC_pct_chng               # Adding new column 'PCT_CHANGE' to UMBC_df

In [43]:
UMBC_df             # checking the updated UMBC_df

UNITID                                   INSTNM  ...  Year  PCT_CHANGE
2151  163268  University of Maryland-Baltimore County  ...  1996         NaN
2049  163268  University of Maryland-Baltimore County  ...  1997         NaN
1982  163268  University of Maryland-Baltimore County  ...  1998         NaN
1955  163268  University of Maryland-Baltimore County  ...  1999         NaN
1926  163268  University of Maryland-Baltimore County  ...  2000         NaN
1915  163268  University of Maryland-Baltimore County  ...  2001        7.65
1893  163268  University of Maryland-Baltimore County  ...  2002        7.65
1873  163268  University of Maryland-Baltimore County  ...  2003       16.13
1830  163268  University of Maryland-Baltimore County  ...  2004        8.55
1819  163268  University of Maryland-Baltimore County  ...  2005        6.23
1799  163268  University of Maryland-Baltimore County  ...  2006        1.20
1779  163268  University of Maryland-Baltimore County  ...  2007        1.00
1767  163268  University of Maryland-Baltimore County  ...  2008        0.83
1745  163268  University of Maryland-Baltimore County  ...  2009        1.05
1710  163268  University of Maryland-Baltimore County  ...  2010        3.37
1694  163268  University of Maryland-Baltimore County  ...  2011        3.23
1660  163268  University of Maryland-Baltimore County  ...  2012        3.14
1629  163268  University of Maryland-Baltimore County  ...  2013        3.11
1610  163268  University of Maryland-Baltimore County  ...  2014        3.14
1555  163268  University of Maryland-Baltimore County  ...  2015        5.99
1510  163268  University of Maryland-Baltimore County  ...  2016        2.34
1478  163268  University of Maryland-Baltimore County  ...  2017        2.25
1415  163268  University of Maryland-Baltimore County  ...  2018        2.26

[23 rows x 8 columns]

In [44]:
# Finaly, we can make the plot tuition growth rate year over year - bar first, then line chart

# (write code here)

fig = px.bar(UMBC_df,
             x = 'Year',
             y = 'PCT_CHANGE',
             title = "Rate of tution fee growth as per the year (Bar Chart)",
             color = 'PCT_CHANGE',
             height = 600)
fig.show()

In [45]:
fig = px.line(UMBC_df,
             x = 'Year',
             y = 'PCT_CHANGE',
             title = "Rate of tution fee growth as per the year (Line Chart)")
             
fig.show()

In [46]:
# <12>
# Since we want to do the same calculation for JHU. 
# Instead of doing it piecemeal as we did for UMBC
# Let's create a function which can be reused for any college.
# This function takes a list of tuitions and return a list of percentage changes year over year
# This function can be used later 

#(Write code here)

# Defining percentage change function

def Percent_change(list):
  percent = []
  for i in range(0, len(list)):
    if i == 0:
      percent.append(np.nan)
    else:
      percent.append(((list[i] - list[i-1]) / list[i-1]) * 100)
  return percent

In [47]:
# <13>
# the dataframe contains many years of data of all U.S. colleges.
# let's just look at JHU
# filter/query the dataframe to retrieve only rows that belong to JHU
# save the JHU data to a new data frame. using a new variable 
# so that the old big data frame is still available for later use.

#(Write code here)

final_df[final_df['INSTNM'].str.contains("Johns Hopkins University")]   # getting rows that have JHU in institute name 

UNITID                    INSTNM STABBR  ...  ADM_RATE  TUITIONFEE_IN  Year
1760  162928  Johns Hopkins University     MD  ...    0.2609        37700.0  2008
1884  162928  Johns Hopkins University     MD  ...    0.3506        27390.0  2002
1823  162928  Johns Hopkins University     MD  ...    0.3071        30140.0  2004
1738  162928  Johns Hopkins University     MD  ...    0.2782        39150.0  2009
1503  162928  Johns Hopkins University     MD  ...    0.1284        50410.0  2016
1906  162928  Johns Hopkins University     MD  ...    0.3431        26210.0  2001
1812  162928  Johns Hopkins University     MD  ...    0.3509        31620.0  2005
1946  162928  Johns Hopkins University     MD  ...       NaN            NaN  1999
1865  162928  Johns Hopkins University     MD  ...    0.3060        28730.0  2003
1917  162928  Johns Hopkins University     MD  ...       NaN        24930.0  2000
1703  162928  Johns Hopkins University     MD  ...    0.2225        40680.0  2010
1548  162928  Johns Hopkins University     MD  ...    0.1412        48710.0  2015
1792  162928  Johns Hopkins University     MD  ...    0.2792        33900.0  2006
1603  162928  Johns Hopkins University     MD  ...    0.1608        47060.0  2014
2039  162928  Johns Hopkins University     MD  ...       NaN            NaN  1997
1687  162928  Johns Hopkins University     MD  ...    0.1940        42280.0  2011
2139  162928  Johns Hopkins University     MD  ...       NaN            NaN  1996
1622  162928  Johns Hopkins University     MD  ...    0.1813        45470.0  2013
1653  162928  Johns Hopkins University     MD  ...    0.1843        43390.0  2012
1410  162928  Johns Hopkins University     MD  ...    0.1148        53740.0  2018
1473  162928  Johns Hopkins University     MD  ...    0.1254        52170.0  2017
1772  162928  Johns Hopkins University     MD  ...    0.2571        35900.0  2007
1973  162928  Johns Hopkins University     MD  ...       NaN            NaN  1998

[23 rows x 7 columns]

In [48]:
# assigning this data to a new dataframe 'JHU_df'

JHU_df = final_df[final_df['INSTNM'].str.contains("Johns Hopkins University")]


In [49]:
JHU_df = JHU_df.sort_values(by="Year")
JHU_df

UNITID                    INSTNM STABBR  ...  ADM_RATE  TUITIONFEE_IN  Year
2139  162928  Johns Hopkins University     MD  ...       NaN            NaN  1996
2039  162928  Johns Hopkins University     MD  ...       NaN            NaN  1997
1973  162928  Johns Hopkins University     MD  ...       NaN            NaN  1998
1946  162928  Johns Hopkins University     MD  ...       NaN            NaN  1999
1917  162928  Johns Hopkins University     MD  ...       NaN        24930.0  2000
1906  162928  Johns Hopkins University     MD  ...    0.3431        26210.0  2001
1884  162928  Johns Hopkins University     MD  ...    0.3506        27390.0  2002
1865  162928  Johns Hopkins University     MD  ...    0.3060        28730.0  2003
1823  162928  Johns Hopkins University     MD  ...    0.3071        30140.0  2004
1812  162928  Johns Hopkins University     MD  ...    0.3509        31620.0  2005
1792  162928  Johns Hopkins University     MD  ...    0.2792        33900.0  2006
1772  162928  Johns Hopkins University     MD  ...    0.2571        35900.0  2007
1760  162928  Johns Hopkins University     MD  ...    0.2609        37700.0  2008
1738  162928  Johns Hopkins University     MD  ...    0.2782        39150.0  2009
1703  162928  Johns Hopkins University     MD  ...    0.2225        40680.0  2010
1687  162928  Johns Hopkins University     MD  ...    0.1940        42280.0  2011
1653  162928  Johns Hopkins University     MD  ...    0.1843        43390.0  2012
1622  162928  Johns Hopkins University     MD  ...    0.1813        45470.0  2013
1603  162928  Johns Hopkins University     MD  ...    0.1608        47060.0  2014
1548  162928  Johns Hopkins University     MD  ...    0.1412        48710.0  2015
1503  162928  Johns Hopkins University     MD  ...    0.1284        50410.0  2016
1473  162928  Johns Hopkins University     MD  ...    0.1254        52170.0  2017
1410  162928  Johns Hopkins University     MD  ...    0.1148        53740.0  2018

[23 rows x 7 columns]

#### Now the dataframe is sorted according to the year in ascending order and it looks good for the analysis.

In [50]:
# <14> 
# Plot JHU's in-state tuition overtime from 1996 to 2019. 
# Let's plot bar and line chart together

#(Write code here)

# Plotting bar and line charts together

fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=JHU_df['Year'],
        y=JHU_df['TUITIONFEE_IN'],
        name="In-state tution fee as per year (Line Chart)"
    ))

fig.add_trace(
    go.Bar(
        x=JHU_df['Year'],
        y=JHU_df['TUITIONFEE_IN'],
        name='In-state tution fee as per year (Bar Chart)'
       
    ))

fig.show()

In [51]:
JHU_fee_list = JHU_df['TUITIONFEE_IN'].tolist()     # Assigning the column 'TUITIONFEE_IN' to a list 'JHU_fee_list'
JHU_fee_list

[nan,
 nan,
 nan,
 nan,
 24930.0,
 26210.0,
 27390.0,
 28730.0,
 30140.0,
 31620.0,
 33900.0,
 35900.0,
 37700.0,
 39150.0,
 40680.0,
 42280.0,
 43390.0,
 45470.0,
 47060.0,
 48710.0,
 50410.0,
 52170.0,
 53740.0]

In [52]:
# <15>
# Let's calculate the JHU tuition percentge growth using the function defined earlier

#(Write code here)

JHU_pct_chng = Percent_change(JHU_fee_list)        # calculating percent growth using pre-defined function
JHU_pct_chng              

[nan,
 nan,
 nan,
 nan,
 nan,
 5.134376253509827,
 4.502098435711561,
 4.8922964585615185,
 4.907761921336582,
 4.91041804910418,
 7.2106261859582546,
 5.899705014749262,
 5.013927576601671,
 3.8461538461538463,
 3.9080459770114944,
 3.933136676499508,
 2.6253547776726585,
 4.793731274487209,
 3.496811084231361,
 3.5061623459413513,
 3.4900431122972697,
 3.4913707597698873,
 3.0093923710944988]

In [53]:
JHU_pct_chng = [round(i, 2) for i in JHU_pct_chng]      # rounding the values to 2 decimals
JHU_pct_chng

[nan,
 nan,
 nan,
 nan,
 nan,
 5.13,
 4.5,
 4.89,
 4.91,
 4.91,
 7.21,
 5.9,
 5.01,
 3.85,
 3.91,
 3.93,
 2.63,
 4.79,
 3.5,
 3.51,
 3.49,
 3.49,
 3.01]

In [54]:
# <16>
# Add the list of percentage changes to the dataframe as a new column ("PCT_CHANGE")

#(Write code here)

JHU_df["PCT_CHANGE"] = JHU_pct_chng

In [55]:
JHU_df                # checking the updated JHU dataframe

UNITID                    INSTNM STABBR  ...  TUITIONFEE_IN  Year  PCT_CHANGE
2139  162928  Johns Hopkins University     MD  ...            NaN  1996         NaN
2039  162928  Johns Hopkins University     MD  ...            NaN  1997         NaN
1973  162928  Johns Hopkins University     MD  ...            NaN  1998         NaN
1946  162928  Johns Hopkins University     MD  ...            NaN  1999         NaN
1917  162928  Johns Hopkins University     MD  ...        24930.0  2000         NaN
1906  162928  Johns Hopkins University     MD  ...        26210.0  2001        5.13
1884  162928  Johns Hopkins University     MD  ...        27390.0  2002        4.50
1865  162928  Johns Hopkins University     MD  ...        28730.0  2003        4.89
1823  162928  Johns Hopkins University     MD  ...        30140.0  2004        4.91
1812  162928  Johns Hopkins University     MD  ...        31620.0  2005        4.91
1792  162928  Johns Hopkins University     MD  ...        33900.0  2006        7.21
1772  162928  Johns Hopkins University     MD  ...        35900.0  2007        5.90
1760  162928  Johns Hopkins University     MD  ...        37700.0  2008        5.01
1738  162928  Johns Hopkins University     MD  ...        39150.0  2009        3.85
1703  162928  Johns Hopkins University     MD  ...        40680.0  2010        3.91
1687  162928  Johns Hopkins University     MD  ...        42280.0  2011        3.93
1653  162928  Johns Hopkins University     MD  ...        43390.0  2012        2.63
1622  162928  Johns Hopkins University     MD  ...        45470.0  2013        4.79
1603  162928  Johns Hopkins University     MD  ...        47060.0  2014        3.50
1548  162928  Johns Hopkins University     MD  ...        48710.0  2015        3.51
1503  162928  Johns Hopkins University     MD  ...        50410.0  2016        3.49
1473  162928  Johns Hopkins University     MD  ...        52170.0  2017        3.49
1410  162928  Johns Hopkins University     MD  ...        53740.0  2018        3.01

[23 rows x 8 columns]

In [56]:
# <17> Make a bar plot of the JHU percentage changes tuitions over time

#(Write code here)

fig = px.bar(JHU_df,
             x = 'Year',
             y = 'PCT_CHANGE',
             title = 'Rate of tution fee change as per the year (Bar plot)',
             color = 'PCT_CHANGE',
             height = 600)

fig.show()

In [57]:
# <18>
# In order to plot both UMBC and JHU tuition change over time in the same plot,
# we need to combine the two datasets using the common key of YEAR 
# First make a umbc2 dataframe with only two columns needed. WE don't need other columns.
# Also change the column name from "PCT_CHANGE" to "UMBC_PCT" in preparation for the merge.
# This is because both umbc and jhu dataframe have the same column name "PCT_CHANGE",
# We rename them to there is no collision during the merge. BTW, Pandas handles collision gracefully,
# Feel free to try it without changing the column names.

#(Write code here)

UMBC_df

UNITID                                   INSTNM  ...  Year  PCT_CHANGE
2151  163268  University of Maryland-Baltimore County  ...  1996         NaN
2049  163268  University of Maryland-Baltimore County  ...  1997         NaN
1982  163268  University of Maryland-Baltimore County  ...  1998         NaN
1955  163268  University of Maryland-Baltimore County  ...  1999         NaN
1926  163268  University of Maryland-Baltimore County  ...  2000         NaN
1915  163268  University of Maryland-Baltimore County  ...  2001        7.65
1893  163268  University of Maryland-Baltimore County  ...  2002        7.65
1873  163268  University of Maryland-Baltimore County  ...  2003       16.13
1830  163268  University of Maryland-Baltimore County  ...  2004        8.55
1819  163268  University of Maryland-Baltimore County  ...  2005        6.23
1799  163268  University of Maryland-Baltimore County  ...  2006        1.20
1779  163268  University of Maryland-Baltimore County  ...  2007        1.00
1767  163268  University of Maryland-Baltimore County  ...  2008        0.83
1745  163268  University of Maryland-Baltimore County  ...  2009        1.05
1710  163268  University of Maryland-Baltimore County  ...  2010        3.37
1694  163268  University of Maryland-Baltimore County  ...  2011        3.23
1660  163268  University of Maryland-Baltimore County  ...  2012        3.14
1629  163268  University of Maryland-Baltimore County  ...  2013        3.11
1610  163268  University of Maryland-Baltimore County  ...  2014        3.14
1555  163268  University of Maryland-Baltimore County  ...  2015        5.99
1510  163268  University of Maryland-Baltimore County  ...  2016        2.34
1478  163268  University of Maryland-Baltimore County  ...  2017        2.25
1415  163268  University of Maryland-Baltimore County  ...  2018        2.26

[23 rows x 8 columns]

In [58]:
UMBC2_df = UMBC_df.iloc[:,[6,7]]          # taking only the required columns
UMBC2_df

Year  PCT_CHANGE
2151  1996         NaN
2049  1997         NaN
1982  1998         NaN
1955  1999         NaN
1926  2000         NaN
1915  2001        7.65
1893  2002        7.65
1873  2003       16.13
1830  2004        8.55
1819  2005        6.23
1799  2006        1.20
1779  2007        1.00
1767  2008        0.83
1745  2009        1.05
1710  2010        3.37
1694  2011        3.23
1660  2012        3.14
1629  2013        3.11
1610  2014        3.14
1555  2015        5.99
1510  2016        2.34
1478  2017        2.25
1415  2018        2.26

In [59]:
UMBC2_df.rename(columns = {'PCT_CHANGE':'UMBC_PCT'}, inplace = True)      # changing the column name to avoid collision during merge

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [60]:
UMBC2_df              # required dataframe for UMBC for merging

Year  UMBC_PCT
2151  1996       NaN
2049  1997       NaN
1982  1998       NaN
1955  1999       NaN
1926  2000       NaN
1915  2001      7.65
1893  2002      7.65
1873  2003     16.13
1830  2004      8.55
1819  2005      6.23
1799  2006      1.20
1779  2007      1.00
1767  2008      0.83
1745  2009      1.05
1710  2010      3.37
1694  2011      3.23
1660  2012      3.14
1629  2013      3.11
1610  2014      3.14
1555  2015      5.99
1510  2016      2.34
1478  2017      2.25
1415  2018      2.26

In [61]:
# Then make a jhu2 dataframe with only two columns needed. WE don't need other columns.
# Also change the column name from "PCT_CHANGE" to "JHU_PCT"  in preparation for the merge

#(Write code here)

JHU_df

UNITID                    INSTNM STABBR  ...  TUITIONFEE_IN  Year  PCT_CHANGE
2139  162928  Johns Hopkins University     MD  ...            NaN  1996         NaN
2039  162928  Johns Hopkins University     MD  ...            NaN  1997         NaN
1973  162928  Johns Hopkins University     MD  ...            NaN  1998         NaN
1946  162928  Johns Hopkins University     MD  ...            NaN  1999         NaN
1917  162928  Johns Hopkins University     MD  ...        24930.0  2000         NaN
1906  162928  Johns Hopkins University     MD  ...        26210.0  2001        5.13
1884  162928  Johns Hopkins University     MD  ...        27390.0  2002        4.50
1865  162928  Johns Hopkins University     MD  ...        28730.0  2003        4.89
1823  162928  Johns Hopkins University     MD  ...        30140.0  2004        4.91
1812  162928  Johns Hopkins University     MD  ...        31620.0  2005        4.91
1792  162928  Johns Hopkins University     MD  ...        33900.0  2006        7.21
1772  162928  Johns Hopkins University     MD  ...        35900.0  2007        5.90
1760  162928  Johns Hopkins University     MD  ...        37700.0  2008        5.01
1738  162928  Johns Hopkins University     MD  ...        39150.0  2009        3.85
1703  162928  Johns Hopkins University     MD  ...        40680.0  2010        3.91
1687  162928  Johns Hopkins University     MD  ...        42280.0  2011        3.93
1653  162928  Johns Hopkins University     MD  ...        43390.0  2012        2.63
1622  162928  Johns Hopkins University     MD  ...        45470.0  2013        4.79
1603  162928  Johns Hopkins University     MD  ...        47060.0  2014        3.50
1548  162928  Johns Hopkins University     MD  ...        48710.0  2015        3.51
1503  162928  Johns Hopkins University     MD  ...        50410.0  2016        3.49
1473  162928  Johns Hopkins University     MD  ...        52170.0  2017        3.49
1410  162928  Johns Hopkins University     MD  ...        53740.0  2018        3.01

[23 rows x 8 columns]

In [62]:
JHU2_df = JHU_df.iloc[:,[6,7]]          # taking out the required columns for merging
JHU2_df

Year  PCT_CHANGE
2139  1996         NaN
2039  1997         NaN
1973  1998         NaN
1946  1999         NaN
1917  2000         NaN
1906  2001        5.13
1884  2002        4.50
1865  2003        4.89
1823  2004        4.91
1812  2005        4.91
1792  2006        7.21
1772  2007        5.90
1760  2008        5.01
1738  2009        3.85
1703  2010        3.91
1687  2011        3.93
1653  2012        2.63
1622  2013        4.79
1603  2014        3.50
1548  2015        3.51
1503  2016        3.49
1473  2017        3.49
1410  2018        3.01

In [63]:
JHU2_df.rename(columns = {'PCT_CHANGE':'JHU_PCT'}, inplace = True)          # replacing the column name to avoid collision during merging

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [64]:
JHU2_df                               # required dataframe for merging

Year  JHU_PCT
2139  1996      NaN
2039  1997      NaN
1973  1998      NaN
1946  1999      NaN
1917  2000      NaN
1906  2001     5.13
1884  2002     4.50
1865  2003     4.89
1823  2004     4.91
1812  2005     4.91
1792  2006     7.21
1772  2007     5.90
1760  2008     5.01
1738  2009     3.85
1703  2010     3.91
1687  2011     3.93
1653  2012     2.63
1622  2013     4.79
1603  2014     3.50
1548  2015     3.51
1503  2016     3.49
1473  2017     3.49
1410  2018     3.01

In [65]:
# Now, merger the two data frames
# and display the new dataframe

#(Write code here)

UMBC_JHU_merged = UMBC2_df.merge(JHU2_df)       # merging both the dataframes
UMBC_JHU_merged                 # displaying the merged dataframe

Year  UMBC_PCT  JHU_PCT
0   1996       NaN      NaN
1   1997       NaN      NaN
2   1998       NaN      NaN
3   1999       NaN      NaN
4   2000       NaN      NaN
5   2001      7.65     5.13
6   2002      7.65     4.50
7   2003     16.13     4.89
8   2004      8.55     4.91
9   2005      6.23     4.91
10  2006      1.20     7.21
11  2007      1.00     5.90
12  2008      0.83     5.01
13  2009      1.05     3.85
14  2010      3.37     3.91
15  2011      3.23     3.93
16  2012      3.14     2.63
17  2013      3.11     4.79
18  2014      3.14     3.50
19  2015      5.99     3.51
20  2016      2.34     3.49
21  2017      2.25     3.49
22  2018      2.26     3.01

In [70]:
# Plot UMBC and JHU separately on the same plot 
# However, this does not work since the bars overlap.

#(Write code here)

fig = px.bar(UMBC_JHU_merged,
             x = 'Year',
             y = 'UMBC_PCT',
             title = 'Tution fee change rate with year for UMBC')

fig = px.bar(UMBC_JHU_merged,
             x = 'Year',
             y = 'JHU_PCT',
             title = 'Tution fee change rate with year for JHU')

fig.show()

In [74]:
# Plot them together using y=["UMBC_PCT","JHU_PCT"] option of bar plot

#(Write code here)

merged_final_df = UMBC_JHU_merged.melt(id_vars = 'Year', value_vars = ['UMBC_PCT', 'JHU_PCT'])

In [75]:
merged_final_df

Year  variable  value
0   1996  UMBC_PCT    NaN
1   1997  UMBC_PCT    NaN
2   1998  UMBC_PCT    NaN
3   1999  UMBC_PCT    NaN
4   2000  UMBC_PCT    NaN
5   2001  UMBC_PCT   7.65
6   2002  UMBC_PCT   7.65
7   2003  UMBC_PCT  16.13
8   2004  UMBC_PCT   8.55
9   2005  UMBC_PCT   6.23
10  2006  UMBC_PCT   1.20
11  2007  UMBC_PCT   1.00
12  2008  UMBC_PCT   0.83
13  2009  UMBC_PCT   1.05
14  2010  UMBC_PCT   3.37
15  2011  UMBC_PCT   3.23
16  2012  UMBC_PCT   3.14
17  2013  UMBC_PCT   3.11
18  2014  UMBC_PCT   3.14
19  2015  UMBC_PCT   5.99
20  2016  UMBC_PCT   2.34
21  2017  UMBC_PCT   2.25
22  2018  UMBC_PCT   2.26
23  1996   JHU_PCT    NaN
24  1997   JHU_PCT    NaN
25  1998   JHU_PCT    NaN
26  1999   JHU_PCT    NaN
27  2000   JHU_PCT    NaN
28  2001   JHU_PCT   5.13
29  2002   JHU_PCT   4.50
30  2003   JHU_PCT   4.89
31  2004   JHU_PCT   4.91
32  2005   JHU_PCT   4.91
33  2006   JHU_PCT   7.21
34  2007   JHU_PCT   5.90
35  2008   JHU_PCT   5.01
36  2009   JHU_PCT   3.85
37  2010   JHU_PCT   3.91
38  2011   JHU_PCT   3.93
39  2012   JHU_PCT   2.63
40  2013   JHU_PCT   4.79
41  2014   JHU_PCT   3.50
42  2015   JHU_PCT   3.51
43  2016   JHU_PCT   3.49
44  2017   JHU_PCT   3.49
45  2018   JHU_PCT   3.01

In [85]:
# plotting bar chart for merged_final_df

fig = px.bar(merged_final_df,
             x = 'Year',
             y = 'value',
             color = 'variable',
             barmode = 'group',
             labels = {
                 "YEAR" : "Year",
                 "value" : "Percentage change in tution",
                 "variable" : "College"},
             title = "Bar chart for Percentage Change in tution fee for UMBC and JHU",
             height = 600
             )

fig.show()

In [84]:
# Plot them together using y=["UMBC_PCT","JHU_PCT"] option of line plot

#(Write code here)

# plotting line chart for merged_final_df

fig = px.line(merged_final_df,
              x = 'Year',
              y = 'value',
              color = 'variable',
              labels = {
                  "YEAR" : "Year",
                  "variable" : "College"},
              title = "Line chart for Percentage Change in tution fee for UMBC and JHU"
              )

fig.show()

In [68]:
# <19> 
# document your observation/conclusion of the comparison plots (bar/line)
# use the following Markdown cell

#### Based on my observation
#### Let us consider Bar chart initially. We can clearly see that there is a huge percentage change of tution fee for the UMBC in the year 2003 with 16.13 % and the least percentage change of tution fee for the UMBC is in the year 2008 with 0.83% 
#### Also we can say that the highest percentage change in the tution fee for the JHU is in the yera 2006 with 7.21% and the least percentage change in tution fee for the JHU is in the year 2012 with 2.63%

#### The same information comes from the line chart as well. But I feel bar chart is visually good for analysis and easy to understand when compared to line chart.

#### The End